### PR TITLE
cannon: Add MTCannon-specific differential tests

### LINKED ISSUE #11655: Add tests for new syscalls

### LINKED ISSUE #11656: Test edgecase where we move into wakeup traversal when the left thread stack is empty

### LINKED ISSUE #11657: Add tests for traversal behavior (normal and wakeup)
Need to add theses for the differential tests.

### LINKED ISSUE #11659: Test edgecase where we pop the last thread due to exit syscall

### PR DESCRIPTION
**Pull request by mbaxter**
**Description**

Add tests to cover new functionality in MTCannon: test new syscalls and thread traversal.  

Other changes include:
- Some fixes:
  - Fix FutexWait - don't set thread.FutexAddr  when the syscall fails
  - Remove duplicate sycall constants
- Update clone fuzz test for MTCannon
- Add `ExpectedMTState` utility for validating full state of MTCannon
- Reorganize some test utils:
  - Move randomization logic into single- and multithreaded testutil packages
  - Break up some testutils across multiple files
- Organize syscall constants more consistently across on- and off-chain implementations

**Metadata**

- Fixes https://github.com/ethereum-optimism/optimism/issues/11655
- Fixes https://github.com/ethereum-optimism/optimism/issues/11656
- Fixes https://github.com/ethereum-optimism/optimism/issues/11659
- Fixes https://github.com/ethereum-optimism/optimism/issues/11657

### CODE REVIEW

File: packages/contracts-bedrock/src/cannon/libraries/MIPSSyscalls.sol [hunk 1]
```diff
@@ -42,8 +42,7 @@ library MIPSSyscalls {
     uint32 internal constant SYS_OPEN = 4005;
     uint32 internal constant SYS_NANOSLEEP = 4166;
     // unused syscalls
-    uint32 internal constant SYS_CLOCK_GETTIME = 4263;
```
Comment (mbaxter) [inline packages/contracts-bedrock/src/cannon/libraries/MIPSSyscalls.sol:45]: Duplicate definition below: `SYS_CLOCKGETTIME`

File: packages/contracts-bedrock/src/cannon/libraries/MIPSSyscalls.sol [hunk 2]
```diff
@@ -42,8 +42,7 @@ library MIPSSyscalls {
...
-    uint32 internal constant SYS_GET_AFFINITY = 4240;
```
Comment (mbaxter) [inline packages/contracts-bedrock/src/cannon/libraries/MIPSSyscalls.sol:46]: Duplicate definition a couple lines below: `SYS_GETAFFINITY `

File: packages/contracts-bedrock/src/cannon/libraries/MIPSSyscalls.sol [hunk 3]
```diff
@@ -69,13 +68,13 @@ library MIPSSyscalls {
...
     uint32 internal constant SYS_MINCORE = 4217;
     uint32 internal constant SYS_TGKILL = 4266;
+
     // profiling-related syscalls - ignored
     uint32 internal constant SYS_SETITIMER = 4104;
     uint32 internal constant SYS_TIMERCREATE = 4257;
     uint32 internal constant SYS_TIMERSETTIME = 4258;
     uint32 internal constant SYS_TIMERDELETE = 4261;
     uint32 internal constant SYS_CLOCKGETTIME = 4263;
-    uint32 internal constant SYS_MUNMAP = 4091;
```
Comment (mbaxter) [inline packages/contracts-bedrock/src/cannon/libraries/MIPSSyscalls.sol:78]: Moved this up to group it with the regular noop syscalls

File: packages/contracts-bedrock/src/cannon/MIPS2.sol [hunk 4]
```diff
@@ -202,7 +202,7 @@ contract MIPS2 is ISemver {
                 }
             }
 
-            if (state.stepsSinceLastContextSwitch == sys.SCHED_QUANTUM) {
+            if (state.stepsSinceLastContextSwitch >= sys.SCHED_QUANTUM) {
```
Comment (mbaxter) [inline packages/contracts-bedrock/src/cannon/MIPS2.sol:205]: The `>` case should never catch, but just feels better to have `>=`  😃  This was previously inconsistent across on- and off-chain impls.

File: cannon/mipsevm/multithreaded/mips.go [hunk 5]
```diff
@@ -102,13 +102,13 @@ func (m *InstrumentedState) handleSyscall() error {
 		// args: a0 = addr, a1 = op, a2 = val, a3 = timeout
 		switch a1 {
 		case exec.FutexWaitPrivate:
-			thread.FutexAddr = a0
```
Comment (mbaxter) [inline cannon/mipsevm/multithreaded/mips.go:105]: Only set this on success

File: cannon/mipsevm/multithreaded/testutil/mutators.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+
+	"github.com/ethereum-optimism/optimism/cannon/mipsevm/exec"
+	"github.com/ethereum-optimism/optimism/cannon/mipsevm/multithreaded"
+	"github.com/ethereum-optimism/optimism/cannon/mipsevm/testutil"
+)
+
+type StateMutatorMultiThreaded struct {
```
Comment (mbaxter) [inline cannon/mipsevm/multithreaded/testutil/mutators.go:14]: The `StateMutatorMultiThreaded` logic was just moved from `testutil/state.go`

File: cannon/mipsevm/multithreaded/testutil/expectations.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+	require.Equal(t, e.Heap, actualState.GetHeap(), fmt.Sprintf("Expect heap = 0x%x", e.Heap))
+	require.Equal(t, e.ExitCode, actualState.GetExitCode(), fmt.Sprintf("Expect exitCode = 0x%x", e.ExitCode))
+	require.Equal(t, e.Exited, actualState.GetExited(), fmt.Sprintf("Expect exited = %v", e.Exited))
+	require.Equal(t, e.Step, actualState.GetStep(), fmt.Sprintf("Expect step = %d", e.Step))
+	require.Equal(t, e.LastHint, actualState.GetLastHint(), fmt.Sprintf("Expect lastHint = %v", e.LastHint))
+	require.Equal(t, e.MemoryRoot, common.Hash(actualState.GetMemory().MerkleRoot()), fmt.Sprintf("Expect memory root = %v", e.MemoryRoot))
+	// Thread-related global fields
+	require.Equal(t, e.StepsSinceLastContextSwitch, actualState.StepsSinceLastContextSwitch, fmt.Sprintf("Expect StepsSinceLastContextSwitch = %v", e.StepsSinceLastContextSwitch))
+	require.Equal(t, e.Wakeup, actualState.Wakeup, fmt.Sprintf("Expect Wakeup = %v", e.Wakeup))
+	require.Equal(t, e.TraverseRight, actualState.TraverseRight, fmt.Sprintf("Expect TraverseRight = %v", e.TraverseRight))
```
Comment (tynes) [inline cannon/mipsevm/multithreaded/testutil/expectations.go]: You can use `require.Equalf` to remove the need for the `fmt.Sprintf`

https://pkg.go.dev/github.com/stretchr/testify/require#Equalf

Comment (mbaxter) [reply cannon/mipsevm/multithreaded/testutil/expectations.go]: awesome - thanks!

**Review by Inphi — APPROVED** [review]
Nice and thorough!

**Pull request merged**